wai-extra/test/Network/Wai/TestSpec.hs

```diff
@@ -223,5 +227,15 @@ spec = do
                    setPath defaultRequest "/get"
       simpleBody sresp `shouldBe` "[]"
 
+  describe "WaiTestFailure" $ do
+    context "when used as a pattern" $ do
+      it "matches HUnitFailure" $ do
+        Left (WaiTestFailure err) <- try $ 23 `shouldBe` (42 :: Int)
+        err `shouldBe` "expected: 42\n but got: 23"
 
+    context "when used as an expression" $ do
+      it "it types as HUnitFailure" $ do
+        throwIO (WaiTestFailure "foo") `shouldThrow` instanceOf (undefined :: HUnitFailure)
 
+      it "it types as WaiTestFailure" $ do
+        throwIO (WaiTestFailure "foo") `shouldThrow` instanceOf (undefined :: WaiTestFailure)
```

Comment:
@snoyberg this is what we get by turning `WaiTestFailure` into a pattern synonym.  It's close to, but not 100% behavior preserving, as catching `WaiTestFailure` will now catch all instances of `HUnitFailure`.

I still hope that this is good enough for all practical purposes and that it provides a smooth upgrade path.